src/operator/contrib/yolo_output-inl.h

```diff
@@ -0,0 +1,703 @@
+    .describe("Predefined anchor box widths and heights.");
+    DMLC_DECLARE_FIELD(warmup_samples).set_default(12800)
+    .describe("Number of images to warm up towards averaging position for box "
+    "predictions when starting a new training. ");
+    DMLC_DECLARE_FIELD(warmup_grad_scale).set_default(0.01)
+    .describe("Gradient scale for non-critical anchors during warm-up stage.");
+    DMLC_DECLARE_FIELD(nms_threshold).set_default(0.5f)
+    .describe("Non-maximum suppression threshold.");
+    DMLC_DECLARE_FIELD(force_suppress).set_default(false)
+    .describe("Suppress all detections regardless of class_id.");
+    DMLC_DECLARE_FIELD(nms_topk).set_default(-1)
+    .describe("Keep maximum top k detections before nms, -1 for no limit.");
+  }
+};  // struct Yolo2OutputParam
+
+template<typename DType>
+MSHADOW_XINLINE DType Intersect(DType l1, DType r1, DType l2, DType r2) {
+  DType left = l1 > l2 ? l1 : l2;
+  DType right = r1 < r2 ? r1 : r2;
+  DType w = right - left;
+  return w > 0 ? w : DType(0);
+}
+
+template<typename DType>
+MSHADOW_XINLINE DType Area(DType l1, DType t1, DType r1, DType b1) {
+  DType width = r1 - l1;
+  DType height = b1 - t1;
+  if (width <= 0 || height <= 0) return DType(0);
+  return width * height;
+}
+
+template<typename DType>
+MSHADOW_XINLINE DType IOU(DType l1, DType t1, DType r1, DType b1,
+  DType l2, DType t2, DType r2, DType b2) {
+  DType inter_area = Intersect(l1, r1, l2, r2) * Intersect(t1, b1, t2, b2);
+  if (inter_area <= 0) return DType(0);
+  DType area1 = Area(l1, t1, r1, b1);
+  DType area2 = Area(l2, t2, r2, b2);
+  return inter_area / (area1 + area2 - inter_area);
+}
+
+// compute intersection-over-union overlap between two boxes
+struct calc_overlap {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out,
+      const DType* L1, const DType* T1, const DType* R1, const DType* B1,
+      const DType* L2, const DType* T2, const DType* R2, const DType* B2) {
+    out[i] = IOU(L1[i], T1[i], R1[i], B1[i], L2[i], T2[i], R2[i], B2[i]);
+  }
+};
+
+struct clip_zero_one {
+  template<typename DType>
+  MSHADOW_XINLINE static DType Map(DType a) {
+    if (a < 0.f) return DType(0.f);
+    if (a > 1.f) return DType(1.f);
+    return DType(a);
+  }
+};  // struct clip_zero_one
+
+
+// find best anchor box per ground-truth, and calculate grad
+struct box_grad {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* grad, DType* out_label,
+      const DType* label, const DType* anchor, const DType* pred,
+      const index_t label_width, const index_t label_offset,
+      const index_t pred_width, const index_t pred_offset,
+      const index_t grad_width, const index_t grad_offset,
+      const index_t num_anchor, const index_t num_label,
+      const index_t width, const index_t height,
+      const float box_scale, const float object_scale) {
+    for (int n = 0; n < static_cast<int>(num_label); ++n) {
```

Comment:
why not iterate with index_t or size_t directly? Safer in the unlikely case of overflow.

Reply:
ints are passed to potentially cuda kernels, which is perferrable than size_t or index_t(defined in mshadow) to avoid troubles, and int is considerably large enough for the purpose of this operator.